spanner/batchwriter.go

```diff
@@ -0,0 +1,320 @@
+	retryLimit int64                      // Limit on retries.
+	verbose    bool                       // If true, print out messages about each write batch.
+	a          asyncState
+}
+
+type row struct {
+	table string
+	cols  []string
+	vals  []interface{}
+}
+
+// Fields in this struct are modified asynchronously e.g. by go routines writing
+// data to Spanner. Either hold a lock or use atomics, as detailed below.
+type asyncState struct {
+	writes       int64            // Number of in-progress writes; access using atomic.
+	retries      int64            // Number of retries; access using atomic.
+	lock         sync.Mutex       // Protects errors and badRows
+	errors       map[string]int64 // Errors encountered; protected by lock.
+	badRows      []*row           // Rows that generated errors; protected by lock.
+	badRowsBytes int64            // Estimate of bytes for badRows; protected by lock.
+	droppedRows  map[string]int64 // Count of dropped rows, broken down by table.
+}
+
+// BatchWriterConfig specifies parameters for configuring BatchWriter.
+type BatchWriterConfig struct {
+	WriteLimit int64                      // Limit on number of in-progress writes.
+	BytesLimit int64                      // Limit on bytes buffered.
+	RetryLimit int64                      // Limit on retries.
+	Write      func([]*sp.Mutation) error // Function to call to write to Spanner (typically a closure that calls client.Apply).
+	Verbose    bool                       // If true, print out messages about each write batch.
+}
+
+// NewBatchWriter returns a new BatchWriter with parameters defined by config.
+func NewBatchWriter(config BatchWriterConfig) *BatchWriter {
+	return &BatchWriter{
+		write:      config.Write,
+		writeLimit: config.WriteLimit,
+		bytesLimit: config.BytesLimit,
+		retryLimit: config.RetryLimit,
+		verbose:    config.Verbose,
+		a: asyncState{
+			errors:      make(map[string]int64),
+			droppedRows: make(map[string]int64),
+		},
+	}
+}
+
+// AddRow appends a new row of data to bw's buffer of rows. Depending on the
+// state of BatchWriter, AddRow may immediately return, or it may initiate writes,
+// or it may block (waiting for some of the writes already in progress to
+// complete) and then initiate writes.
+func (bw *BatchWriter) AddRow(table string, cols []string, vals []interface{}) {
+	r := &row{table, cols, vals}
+	bw.rows = append(bw.rows, r)
+	bw.rBytes += byteSize(r)
+	bw.rCount += int64(len(r.cols))
+	bw.writeData()
+}
+
+// Flush initiates writes to Spanner of all buffered rows of data, and waits
+// for them to complete.
+func (bw *BatchWriter) Flush() {
+	for len(bw.rows) > 0 {
+		if atomic.LoadInt64(&bw.a.writes) < bw.writeLimit {
+			m, count, bytes := bw.getBatch()
+			if bw.verbose {
+				fmt.Printf("Starting write of %d rows to Spanner (%d bytes, %d mutations) [%d in progress]\n",
+					len(m), bytes, count, atomic.LoadInt64(&bw.a.writes))
+			}
+			bw.startWrite(m)
+		} else {
+			time.Sleep(10 * time.Millisecond)
```

Comment:
Flush needs to wait till completion in any case.

The sleep here (and the one in writeData) are because we don't want to have too many pending writes out to Spanner. The main reason for this is that if we just churned through the pg_dump file and generated lots of pending mutations all waiting to be written to Spanner, we'd consume considerable memory in harbourbridge and we'd OOM on larger pg_dumps.

To make harbourbridge more robust and limit its memory footprint, we limit the number of pending Spanner writes and the amount of data we keep in rows, and if we exceed these limits, we throttle our processing of the pg_dump i.e. we leave the data on-disk until we have capacity to process it.

Reply:
Completely agree that it's a good idea to limit the memory usage.

Instead of "polling" `bw.a.writes`, could this be implemented through channels?

Reply:
As discussed, let's defer for now, and I'll add a feature request to make this whole library thread-safe, and (probably) use a channel for submitting rows to batchwriter.